votor/src/root_utils.rs

```diff
@@ -0,0 +1,142 @@
+    drop_bank_sender: &Sender<Vec<BankWithScheduler>>,
+    blockstore: &Blockstore,
+    leader_schedule_cache: &Arc<LeaderScheduleCache>,
+    bank_forks: &RwLock<BankForks>,
+    rpc_subscriptions: Option<&RpcSubscriptions>,
+    my_pubkey: &Pubkey,
+    callback: CB,
+) -> Result<(), SetRootError>
+where
+    CB: FnOnce(&BankForks),
+{
+    // get the root bank before squash
+    let root_bank = bank_forks
+        .read()
+        .unwrap()
+        .get(new_root)
+        .expect("Root bank doesn't exist");
+    let mut rooted_banks = root_bank.parents();
+    let oldest_parent = rooted_banks.last().map(|last| last.parent_slot());
+    rooted_banks.push(root_bank.clone());
+    let rooted_slots: Vec<_> = rooted_banks.iter().map(|bank| bank.slot()).collect();
+    // The following differs from rooted_slots by including the parent slot of the oldest parent bank.
+    let rooted_slots_with_parents = bank_notification_sender
+        .as_ref()
+        .is_some_and(|sender| sender.should_send_parents)
+        .then(|| {
+            let mut new_chain = rooted_slots.clone();
+            new_chain.push(oldest_parent.unwrap_or(parent_slot));
+            new_chain
+        });
+
+    // Call leader schedule_cache.set_root() before blockstore.set_root() because
+    // bank_forks.root is consumed by repair_service to update gossip, so we don't want to
+    // get shreds for repair on gossip before we update leader schedule, otherwise they may
+    // get dropped.
+    leader_schedule_cache.set_root(rooted_banks.last().unwrap());
+    blockstore
+        .set_roots(rooted_slots.iter())
+        .expect("Ledger set roots failed");
+    set_bank_forks_root(
+        new_root,
+        bank_forks,
+        snapshot_controller,
+        highest_super_majority_root,
+        drop_bank_sender,
+        callback,
+    )?;
+    blockstore.slots_stats.mark_rooted(new_root);
+    if let Some(rpc_subscriptions) = rpc_subscriptions {
+        rpc_subscriptions.notify_roots(rooted_slots);
+    }
+    if let Some(sender) = bank_notification_sender {
+        let dependency_work = sender
+            .dependency_tracker
+            .as_ref()
+            .map(|s| s.get_current_declared_work());
+        sender
+            .sender
+            .send((BankNotification::NewRootBank(root_bank), dependency_work))
+            .unwrap_or_else(|err| warn!("bank_notification_sender failed: {err:?}"));
+
+        if let Some(new_chain) = rooted_slots_with_parents {
+            let dependency_work = sender
+                .dependency_tracker
+                .as_ref()
+                .map(|s| s.get_current_declared_work());
+            sender
+                .sender
+                .send((BankNotification::NewRootedChain(new_chain), dependency_work))
+                .unwrap_or_else(|err| warn!("bank_notification_sender failed: {err:?}"));
+        }
+    }
+    info!("{} new root {}", my_pubkey, new_root);
```

Comment:
To confirm, this is the log line you were referencing above ? 

nit: Please change this to put the arguments inline. There is a clippy lint in Rust 1.88 that requires this; we haven't upgraded to 1.88 due to a perf regression. But, I already cleaned up this crate in https://github.com/anza-xyz/agave/pull/6933

Note that I will eventually turn the lint on to prevent the reintroduction of items that break while we wait for a solution in upstream Rust, but I haven't been prioritizing that since we know that 1..89 (the latest stable) doesn't address the per regression either. So, this will save me or someone else a subsequent PR to clean it up 😄